### ISSUE TITLE
A problem about attention mask!

### ISSUE THREAD

**Issue by LindgeW** [andyljones/simple-transformer-xl#1]
_assert M == self.mem_len
all_ones = h.new_ones((T, M+T), dtype=torch.uint8)
mask = (torch.triu(all_ones, M+1) + torch.tril(all_ones, 0))[:, :, None]_ 

if M is zero, mask will be all-one matrix, which means all elements will be filled into _-inf_ after  softmax. Can you interpret it ??

**Comment by andyljones** [comment]
If M is zero, the transformer doesn't have any memory. Deciding what it means to attend to an empty memory is a bit of an edge case. Running a test though, I get uniform distribution over memory cells for `M == 0`, rather than `-inf`?

```python
import torch
from torch.nn import functional as F

T = 3
M = 0

all_ones = torch.ones((T, M+T), dtype=torch.uint8)
mask = (torch.triu(all_ones, M+1) + torch.tril(all_ones, 0))[:, :, None]

attn_score = torch.rand((T, T+M, 1, 1))
attn_score = (attn_score
                .float()
                # -65k is a very, very small number for a 16-bit float
                .masked_fill(mask.bool()[:, :, :, None], -65000)
                .type_as(attn_score))

attn_prob = F.softmax(attn_score, dim=1)
```

That certainly isn't correct behaviour either mind you - it basically lets the transformer 'see into the future'. I don't want to complicate the code to fix this special case, but I will add an assertion warning people about this. Thanks for making me aware of it!

**Comment by LindgeW** [comment]
Yeah! you are right. I made an empirical mistake.

I have read the official code [mem_transformer.py#L649](https://github.com/kimiyoung/transformer-xl/blob/44781ed21dbaec88b280f74d9ae2877f52b492a5/pytorch/mem_transformer.py#L649), and it seems that the attention masks are different between train phase and eval phase.  What your code shows is actually the eval one. 
for eval (self.same_length is True):
`att_mask = (torch.triu(all_ones, M+1) + torch.tril(all_ones, 0))[:, :, None]`

for train (self.same_length is False):
`att_mask = torch.triu(all_ones, M+1)[:, :, None]`

**Comment by andyljones** [comment]
Aha, I understand. OK: `same_length` is not quite equivalent to `eval mode`, it's more that if `same_length` is true, then evaluating the network one step at a time gives the same results as evaluating the network in larger chunks. If it's false, that's not true. 

The downside to `same_length = False` is that you're throwing away of bit of perfectly useful computation, and so (I think) NLP guys prefer to set it to false in training. Thing is, I put this repo together for reinforcement learning rather than NLP, and for RL it's really important that evaluating one step at a time (on the actor) gives the same result as evaluating in whole chunks (on the learner). So for my purposes `same_length` is always true, and so I ignored the other code path.
